### ISSUE TITLE
Refactoring `rust-bert` to have network- and filesystem-access feature gated

### ISSUE THREAD

**Issue by JonasHedEng** [guillaume-be/rust-bert#222]
First off, thanks for this crate, being able to use Rust throughout our whole NLP backend/service is so much nicer than having to somehow interface with PyTorch.

`Resource`-s in `rust-bert` might refer to remote resources that would trigger the loading of a potentially large resource that would take lots of time to download and use up a lot of disk space.
In a production system, granting access to arbitrary network calls and read/write access to the filesystem are quite significant causes for concern, primarily in regards to security but could also impact performance.
For this reason, putting the convenience functionality of the whole "fetching and caching remote models" behind a feature flag would make `rust-bert` much more viable as a library to use in a production environment.

I've started looking into a refactoring, exploring how this could be done and I'll happily put up a PR if/when it turns out promising, however I'm expecting it to be quite big since `Resource` is such a widely used type.
In the mean time it would be great to get your thoughts on this. Thanks!

**Comment by guillaume-be** [comment]
Hello @JonasHedEng and thank you for your interest in this project!
I agree with you that `RemoteResources` should not be used in a production environment as they query the endpoint to compare the `ETag` and see if the file should be updated - causing the code to fail if a connection cannot be established.

Resources can also be created from [LocalResources](https://docs.rs/rust-bert/latest/rust_bert/resources/struct.LocalResource.html) that essentially just point to a file in the local filesystem. This avoids making any call to the network and relies on other services to ensure the file is available. While I added the remote resources for the convenience of being able to try things out quickly with the library, I would like to avoid making the utilities linked to the download and caching of local files too complex -- as these could easily be delegated to methods outside of the scope of this crate. I was hoping the concept of `LocalResource` would offer flexibility to implement any kind of external mechanism for fetching resource files.

**Comment by JonasHedEng** [comment]
Sorry, I believe I was a bit in-explicit in my issue description.

`LocalResource` works fine and that is the solution I'm intending to use. My concern is rather about still compiling the code and the dependencies for `RemoteResource` since they might trigger external calls such as to the network or files ystem.

We're using [`cap-std`](https://crates.io/crates/cap-std) for anything IO, helping us with security regarding what can be accessed externally.

In the best scenario, even `LocalResource` wouldn't contain the path but rather something that implements `std::io::Read` to fully limit how external access is handled by the caller and not allowing libraries such capabilities at all. Obviously `rust-bert` is just providing the path to [`tch`](https://crates.io/crates/tch) so it's not really a concern for this issue, but hopefully could convey the idea.

Basically I would like to avoid external calls being compiled at all, both for the sake of security and knowing that nothing in the crate could perform such a call(if not enabled by a feature flag), but also to reduce compilation time.
If, in this case, `RemoteResource` can be feature gated completely, neither `cached-path`, `dirs` nor their sub-deps would have to be compiled at all.

Maybe I could put up a Draft PR when I have an initial approach, to better convey my thinking and approach?

**Comment by guillaume-be** [comment]
Hi @JonasHedEng ,

This makes sense thank you for the clarification. I would be interested to see a related PR. Would there be a way to make the feature enabling remote resources turned on by default - making it easier for user to discover and get started with the crate?

**Comment by guillaume-be** [comment]
Thank you very much - I may take a couple of days to review this. I had a quick look at it looks pretty neat! I saw that you used dynamic dispatch and traits for the Resources - meaning one needs to store the resource in a Box before passing them to configs and others. Since all objects are internal I thought one could also go with generics/monomorphism (as this is heavily used throughout the crate) - but after a quick tryout it seems things blow out quickly and one ends up have generics all over the place.  The performance impact of runtime dynamic dispatch is probably negligible anyway (reading resource is done once), but I was wondering if one could save the "boxing" for the end-user. I believe not - but I wanted to double check with you.

I'll get back to you in the next few days after having a closer look - thank you!

**Comment by JonasHedEng** [comment]
No stress! Glad that it's appreciated :)
Yeah, I don't think it would be worth it with generics/monomorphism here. What I'll do though, now that I have gotten it to all compile without `RemoteResource`, is to not have the constructors take a `Box` but rather have generics there that will simply be wrapped by a `Box` when stored in the corresponding `Config` struct. That should make it more convenient and ergonomic to use in most cases!

**Comment by JonasHedEng** [comment]
So I changed it to only have `Box` for the fields and to let the constructors take a generic as parameter and then wrap it in a box.
Everything seems to be compiling and all tests are passing, so I'll take it out of draft status! I hope I've updated the relevant documentation somewhat as well.

**Comment by guillaume-be** [comment]
Fixed by https://github.com/guillaume-be/rust-bert/pull/223
